loader/src/main/java/net/neoforged/fml/ModLoader.java

```diff
@@ -191,73 +190,108 @@ public void gatherAndInitializeMods(final ModWorkManager.DrivenExecutor syncExec
+        var progress = StartupMessageManager.addProgressBar(name, 0);
+        try {
+            periodicTask.run();
+            waitForFuture(name, periodicTask, future);
+        } finally {
+            progress.complete();
         }
+    }
+
+    /**
+     * Dispatches a task across all mod containers in parallel, with progress displayed on the loading screen.
+     */
+    public void dispatchParallelTask(String name, Executor parallelExecutor, Runnable periodicTask, Consumer<ModContainer> task) {
+        var progress = StartupMessageManager.addProgressBar(name, modList.size());
         try {
-            transition.join();
-        } catch (CompletionException e) {
-            loadingStateValid = false;
-            Throwable t = e.getCause();
-            final List<Throwable> notModLoading = Arrays.stream(t.getSuppressed())
-                    .filter(obj -> !(obj instanceof ModLoadingException))
-                    .collect(Collectors.toList());
-            if (!notModLoading.isEmpty()) {
-                LOGGER.fatal("Encountered non-modloading exceptions!", e);
-                statusConsumer.ifPresent(c -> c.accept("ERROR DURING MOD LOADING"));
-                throw e;
-            }
+            periodicTask.run();
+            var futureList = modList.getSortedMods().stream()
+                    .map(modContainer -> {
+                        return CompletableFuture.runAsync(() -> {
+                            ModLoadingContext.get().setActiveContainer(modContainer);
+                            task.accept(modContainer);
+                        }, parallelExecutor).whenComplete((result, exception) -> {
+                            progress.increment();
+                            ModLoadingContext.get().setActiveContainer(null);
+                        });
+                    })
+                    .toList();
+            var singleFuture = ModList.gather(futureList)
+                    .thenCompose(ModList::completableFutureFromExceptionList);
+            waitForFuture(name, periodicTask, singleFuture);
+        } finally {
+            progress.complete();
+        }
+    }
 
-            final List<ModLoadingException> modLoadingExceptions = Arrays.stream(t.getSuppressed())
-                    .filter(ModLoadingException.class::isInstance)
-                    .map(ModLoadingException.class::cast)
-                    .collect(Collectors.toList());
-            LOGGER.fatal(LOADING, "Failed to complete lifecycle event {}, {} errors found", state.name(), modLoadingExceptions.size());
-            statusConsumer.ifPresent(c -> c.accept("ERROR DURING MOD LOADING"));
-            throw new LoadingFailedException(modLoadingExceptions);
+    private void waitForFuture(String name, Runnable ticker, CompletableFuture<?> future) {
+        while (true) {
+            ticker.run();
+            try {
+                future.get(50, TimeUnit.MILLISECONDS);
```

Comment:
Is it guaranteed that all futures being resolved in this method are expected to take less than 50ms?  Seems like a fairly short timeout, but I don't know what's going through here.

Reply:
The idea seems to be to use a short timeout so that the loop will go around and execute the ticker again every 50ms. However, I'm not sure throwing a `TimeoutException` every 50ms is ideal. Constructing exceptions is not fast.

Reply:
The ticker is what does the loading screen re-draw. This controls the FPS of the loading screen basically. Exceptions are not cheap but this should be fine.

Reply:
I know - the goal is just to have _something_. The background waiter does the same. I'm happy to change it in a future PR.